Syntax error in sql file makes cockroach panic

**BUG REPORT**

Please describe the issue you observed, and any steps we can take to reproduce it:

A syntax error in a file containing SQL definitions caused cockroach to panic. I can reproduce it with:

echo ? | cockroach sql --insecure

- Which version of CockroachDB are you using?

$ cockroach version
Build Tag:    v2.0.4
Build Time:   2018/07/17 14:01:56
Distribution: CCL
Platform:     darwin amd64 (x86_64-apple-darwin17.7.0)
Go Version:   go1.10.3
C Compiler:   4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)
Build SHA-1:  d7c99735249c963177eadbc5fd57fc2094de6823
Build Type:   development

- What did you do?

I executed the commands below in the shell:

echo ? | cockroach sql --insecure

- What did you expect to see?

An error message.

- What did you see instead?

A panic:

E180803 13:48:04.494316 1 util/log/crash_reporting.go:185  a panic has occurred!
panic: runtime error: index out of range [recovered]
        panic: runtime error: index out of range

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.RecoverAndReportPanic(0x6145ce0, 0xc42003e080, 0xc4206c4000)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:100 +0x9f
panic(0x5b603c0, 0x7305630)
        /usr/local/Cellar/go/1.10.3/libexec/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/vendor/github.com/knz/go-libedit/unix.libedit_unix.EditLine.Stdout(...)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/vendor/github.com/knz/go-libedit/unix/editline_unix.go:311
github.com/cockroachdb/cockroach/pkg/cli.(*cliState).doPrepareStatementLine(0xc420811520, 0x2, 0x4, 0xa, 0xb, 0x1000)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:919 +0x4f9
github.com/cockroachdb/cockroach/pkg/cli.runInteractive(0xc420520400, 0x0, 0x0)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:1109 +0x201
github.com/cockroachdb/cockroach/pkg/cli.runTerm(0x7328740, 0xc42006fe70, 0x0, 0x1, 0x0, 0x0)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:1179 +0x159
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x7328740, 0xc42006fe70, 0x0, 0x1, 0x0, 0x0)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:40 +0x5a
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).execute(0x7328740, 0xc42006fe30, 0x1, 0x1, 0x7328740, 0xc42006fe30)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:698 +0x46d
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x732afa0, 0x8, 0x0, 0xc42084dee8)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:783 +0x2e4
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).Execute(0x732afa0, 0x18, 0xc42084df00)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:736 +0x2b
github.com/cockroachdb/cockroach/pkg/cli.Run(0xc42003a1c0, 0x2, 0x2, 0xc42003e080, 0xc4206c4000)
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:156 +0x6d
github.com/cockroachdb/cockroach/pkg/cli.Main()
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:51 +0x15d
main.main()
        /private/tmp/cockroach-20180717-53663-1pyrkos/cockroach-v2.0.4/src/github.com/cockroachdb/cockroach/main.go:27 +0x20

- What was the impact?

Some time used finding out what exactly caused the problem in the original sql file.

--- a/pkg/cli/interactive_tests/test_contextual_help.tcl
+++ b/pkg/cli/interactive_tests/test_contextual_help.tcl
@@ -137,4 +137,16 @@ end_test
 interrupt
 eexpect eof
 
+start_test "Check that the hint for a single ? is also printed in non-interactive sessions."
+spawn /bin/bash
+
+send "echo '?' | $argv sql\r"
+eexpect "Note:"
+eexpect JSON
+eexpect "use '??'"
+
+send "exit\r"
+eexpect eof
+end_test
+
 stop_server $argv

--- a/pkg/cli/sql.go
+++ b/pkg/cli/sql.go
@@ -916,7 +916,11 @@ func (c *cliState) doPrepareStatementLine(
 
 	if !endOfStmt {
 		if lastTok == '?' {
-			fmt.Fprintf(c.ins.Stdout(),
+			stdout := os.Stdout
+			if c.hasEditor() {
+				stdout = c.ins.Stdout()
+			}
+			fmt.Fprintf(stdout,
 				"Note: a single '?' is a JSON operator. If you want contextual help, use '??'.\n")
 		}
 		return contState

